src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -99,28 +156,147 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
             }
             catch (Exception ex)
             {
-                // The job has encountered an error it cannot recover from.
-                // Try to update the job to failed state.
-                _logger.LogError(ex, "Encountered an unhandled exception. The job will be marked as failed.");
-
                 _reindexJobRecord.Error.Add(new OperationOutcomeIssue(
                     OperationOutcomeConstants.IssueSeverity.Error,
                     OperationOutcomeConstants.IssueType.Exception,
                     ex.Message));
-                await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
+
+                _reindexJobRecord.FailureCount++;
+
+                _logger.LogError(ex, $"Encountered an unhandled exception. The job failure count increased to {_reindexJobRecord.FailureCount}.");
+
+                await UpdateJobAsync(cancellationToken);
+
+                if (_reindexJobRecord.FailureCount >= _reindexJobConfiguration.ConsecutiveFailuresThreshold)
+                {
+                    await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
+                }
+            }
+        }
+
+        private async Task ProcessQueryAsync(ReindexJobQueryStatus query, CancellationToken cancellationToken)
+        {
+            try
+            {
+                query.Status = OperationStatus.Running;
+                query.LastModified = DateTimeOffset.UtcNow;
+
+                // Query first batch of resources
+                var results = await ExecuteReindexQueryAsync(query, false, cancellationToken);
+
+                // if continuation token then update next query
+                if (!string.IsNullOrEmpty(results.ContinuationToken))
+                {
+                    var nextQuery = new ReindexJobQueryStatus(results.ContinuationToken);
+                    nextQuery.LastModified = DateTimeOffset.UtcNow;
+                    nextQuery.Status = OperationStatus.Queued;
+                    _reindexJobRecord.QueryList.Add(nextQuery);
+                }
+
+                await UpdateJobAsync(cancellationToken);
```

Comment:
When we update the job here, the query status is set to `Running` and that information will be saved as part of the job record. Let's assume the process restarts here and we have to resume our re-index job. This query has not been completed yet. But we will never pick this up again since we are looking only for "Queued" queries in the `ExecuteAsync` code

Reply:
Good catch!  Looking at the code more carefully with this in mind, I think we have a similar problem in other places if the process is shutdown before we are able to complete some processing.  We do have a separate task: 73390 which is focused on how to handle a restart.  I think we will need to check timestamps on the queries and the reindexjob to see if they have become stale and then reset the status.